src/message/moveMessage.js

```diff
@@ -0,0 +1,225 @@
+
+const inputMarginPadding = {
+  paddingHorizontal: 8,
+  paddingVertical: Platform.select({
+    ios: 8,
+    android: 2,
+  }),
+};
+
+export default function MoveMessage(props: Props): Node {
+  const themeContext = useContext(ThemeContext);
+  const backgroundColor = themeContext.backgroundColor;
+  const cardColor = themeContext.cardColor;
+  const auth = useSelector(getAuth);
+  const streams = useSelector(getStreams);
+  const isadmin = props.route.params.isadmin;
+  const id = props.route.params.message.id;
+  const names = [];
+  const [subject, setSubject] = useState(props.route.params.message.subject);
+  const [propagate_mode, setPropagateMode] = useState('javachange_one');
+  const [stream, setStream] = useState(streamNameOfNarrow(props.route.params.messageNarrow));
+  const [topicFocus, setTopicFocus] = useState(false);
+  const messageInputRef = React.createRef<$FlowFixMe>();
+  const topicInputRef = React.createRef<$FlowFixMe>();
+
+  streams.map(item => names.push({ value: item.name }));
+
+  const styles = {
+    parent: {
+      backgroundColor: cardColor,
+    },
+    layout: {
+      margin: 10,
+    },
+    title: {
+      fontSize: 18,
+      color: backgroundColor === 'white' ? 'black' : 'white',
+    },
+    autocompleteWrapper: {
+      position: 'absolute',
+      top: 0,
+      width: '100%',
+    },
+    composeText: {
+      flex: 1,
+      paddingVertical: 8,
+    },
+    topicInput: {
+      height: 50,
+      backgroundColor,
+      ...inputMarginPadding,
+    },
+    viewTitle: {
+      display: 'flex',
+      flexDirection: 'row',
+      justifyContent: 'space-between',
+      alignItems: 'center',
+      height: 50,
+      paddingHorizontal: 10,
+      marginBottom: 20,
+    },
+    textColor: {
+      color: backgroundColor === 'white' ? 'black' : 'white',
+    },
+    picker: { backgroundColor, marginBottom: 20 },
+    submitButton: {
+      marginTop: 10,
+      paddingTop: 15,
+      paddingBottom: 15,
+      marginLeft: 30,
+      marginRight: 30,
+      backgroundColor: BRAND_COLOR,
+      borderRadius: 10,
+      borderWidth: 1,
+    },
+  };
+
+  const updateTextInput = (textInput, text) => {
+    if (textInput === null) {
+      // Depending on the lifecycle events this function is called from,
+      // this might not be set yet.
+      return;
+    }
+
+    // `textInput` is untyped; see definition.
+    textInput.setNativeProps({ text });
+  };
+
+  const handleTopicChange = (Topic: string) => {
+    setSubject(Topic);
+  };
+
+  const handleTopicFocus = () => {
+    setTopicFocus(true);
+  };
+
+  const setTopicInputValue = (Topic: string) => {
+    updateTextInput(topicInputRef.current, subject);
+    handleTopicChange(Topic);
+    setTopicFocus(false);
+  };
+
+  const handleTopicAutocomplete = (Topic: string) => {
+    setTopicInputValue(Topic);
+    messageInputRef.current?.focus();
+  };
+
+  const updateMessage = async () => {
+    const stream_info = await api.getStreamId(auth, stream);
```

Comment:
This will go and make a request to the server to look up what stream has the given name.

We don't want to do that. We already have data on all the streams the user has access to know about, and that's the data we used for presenting the UI the user will have used to select the stream. We should use that data instead.

That's better because not only does it make things faster, and more reliable on a weak network connection -- because we save a request round-trip -- but also it keeps the action we actually take aligned with the action the user chose, eliminating a class of bugs where there was a mismatch between the data we had at the moment the user chose something in the UI and the data the server has by the time it handles the request.

For example, if I go to move something to stream `#awesome-team`, and right around that moment someone else goes and renames `#awesome-team` to `#awesome team`, things should just work -- the messages should get moved to that stream, even though its name was `#awesome team` by the time the request made it to the server.

(This connects to the point above about using stream IDs rather than stream names.)